blockdag/dag.go

```diff
@@ -605,93 +605,52 @@ func (dag *BlockDAG) connectBlock(node *blockNode,
 	return chainUpdates, nil
 }
 
-// calcMultiset returns the multiset of the UTXO of the given block with the given transactions.
-func (node *blockNode) calcMultiset(dag *BlockDAG, transactions []*util.Tx, acceptanceData MultiBlockTxsAcceptanceData, selectedParentUTXO, pastUTXO UTXOSet) (*secp256k1.MultiSet, error) {
-	ms, err := node.pastUTXOMultiSet(dag, acceptanceData, selectedParentUTXO)
-	if err != nil {
-		return nil, err
-	}
+// calcMultiset returns the multiset of the UTXO of the given block.
+func (node *blockNode) calcMultiset(dag *BlockDAG, acceptanceData MultiBlockTxsAcceptanceData,
+	selectedParentPastUTXO UTXOSet) (*secp256k1.MultiSet, error) {
 
-	for _, tx := range transactions {
-		ms, err = addTxToMultiset(ms, tx.MsgTx(), pastUTXO, UnacceptedBlueScore)
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	return ms, nil
+	return node.pastUTXOMultiSet(dag, acceptanceData, selectedParentPastUTXO)
 }
 
-// acceptedSelectedParentMultiset takes the multiset of the selected
-// parent, replaces all the selected parent outputs' blue score with
-// the block blue score and returns the result.
-func (node *blockNode) acceptedSelectedParentMultiset(dag *BlockDAG,
-	acceptanceData MultiBlockTxsAcceptanceData) (*secp256k1.MultiSet, error) {
+func (node *blockNode) pastUTXOMultiSet(dag *BlockDAG, acceptanceData MultiBlockTxsAcceptanceData,
+	selectedParentPastUTXO UTXOSet) (*secp256k1.MultiSet, error) {
 
-	if node.isGenesis() {
-		return secp256k1.NewMultiset(), nil
-	}
-
-	ms, err := dag.multisetStore.multisetByBlockNode(node.selectedParent)
+	ms, err := node.selectedParentMultiset(dag)
 	if err != nil {
 		return nil, err
 	}
 
-	selectedParentAcceptanceData, exists := acceptanceData.FindAcceptanceData(node.selectedParent.hash)
-	if !exists {
-		return nil, errors.Errorf("couldn't find selected parent acceptance data for block %s", node)
-	}
-	for _, txAcceptanceData := range selectedParentAcceptanceData.TxAcceptanceData {
-		tx := txAcceptanceData.Tx
-		msgTx := tx.MsgTx()
-		isCoinbase := tx.IsCoinBase()
-		for i, txOut := range msgTx.TxOut {
-			outpoint := *wire.NewOutpoint(tx.ID(), uint32(i))
+	for _, blockAcceptanceData := range acceptanceData {
+		for _, txAcceptanceData := range blockAcceptanceData.TxAcceptanceData {
+			if !txAcceptanceData.IsAccepted {
+				continue
+			}
 
-			unacceptedEntry := NewUTXOEntry(txOut, isCoinbase, UnacceptedBlueScore)
-			acceptedEntry := NewUTXOEntry(txOut, isCoinbase, node.blueScore)
+			tx := txAcceptanceData.Tx.MsgTx()
 
 			var err error
-			ms, err = removeUTXOFromMultiset(ms, unacceptedEntry, &outpoint)
-			if err != nil {
-				return nil, err
-			}
-
-			ms, err = addUTXOToMultiset(ms, acceptedEntry, &outpoint)
+			ms, err = addTxToMultiset(ms, tx, selectedParentPastUTXO, node.blueScore)
```

Comment:
Please make sure (preferably with a unit-test) that addTxToMultiset doesn't change `node.selectedParentMultiset` in-place in such a way that it affects the MultiSet store.
TBH, I think it does (but maybe there's a copy somewhere in secp256k)